requirements.txt

```diff
@@ -3,7 +3,7 @@
 # Once an enhancement has been merged, the package needs to be added to the installer.sh, the CI Docker images and issues should be raised on
 # keylime/keylime-rpm keylime/keylime-deb and the keylime ansible installer
 
-cryptography>=3.3.2 # BSD/Apache-2.0
+cryptography>=38.0.0 # BSD/Apache-2.0
```

Comment:
With this we would loose CentOS9 and RHEL 9, which are on v36.0: https://rpmfind.net/linux/rpm2html/search.php?query=python3-cryptography&submit=Search+...

Which functions require v38.0?

Reply:
The functions that require it have been moved to cert_utils:
```   
    # NEEDS CRYPTO >= 38.0.0
    idevid_pub = idevid_cert_509.public_key()
    iak_pub = iak_cert_509.public_key()
```
I think CentOS9 and RHEL should both be able to install using the installer as pip will provide a wheel.

Reply:
You can use this to reproduce:
https://gist.github.com/Isaac-Matthews/1ec57e5536b53ecc2349115193c7416c

Reply:
> I think CentOS9 and RHEL should both be able to install using the installer as pip will provide a wheel.

SuSE is even only on 3.3.2

The issue is that keylime from rpm packages is supposed to work right away with what is available.

@mpeters , @aplanas : Any chance that at least SuSE and RHEL can update to a more recent python3-cryptography?

Reply:
@stefanberger Probably not soon, I'll ask the responsible team.

Maybe that is not a problem for now as in CentOS/RHEL keylime is still based on version 7.3.0 (and will be at least for the next 6 months).

> I think CentOS9 and RHEL should both be able to install using the installer as pip will provide a wheel.

Probably installing directly from pip is vetoed in most enterprises. For RHEL, all the dependencies must also be packaged as rpms, meaning that we cannot include these changes in keylime unless the python-cryptography package is updated.

Reply:
SLE / Leap is in 3.3.2 and will not change in the 15.X live cycle. But in MicroOS / Tumbleweed we have 41.0.4. For SUSE we have already a Keylime release, so in that regard is in maintenance mode, so this change is OK.

Reply:
Is it possible to use the `%pyproject_wheel` macro to build and install keylime as a wheel, with the dependencies collected via `%pyproject_buildrequires` as described [here?](https://src.fedoraproject.org/rpms/pyproject-rpm-macros/)

I do not have a great deal of understanding for rpms/packaging so that may not make sense.

Reply:
I'll try to explain without going into RPM details: we cannot use dependencies from pip in CentOS/RHEL for security reasons, not because of technical limitations.

Which means that if we include the changes as they are now and python cryptography is not updated in CentOS/RHEL, we cannot build keylime either. We would be stuck in the current release until python-cryptography is updated, which might mean forever for RHEL 9.

I'd suggest to make the feature enabled conditionally, depending on the available python-cryptography version instead of changing the keylime python-cryptography requirement unconditionally.

Reply:
OK understood. In that case would we be happy with:
- Checking crypto version before idevid chunk and setting tpm_identity as ek_cert so the idevid chunk is never used, maybe putting out a warning if the config option gets changed by this event
- Including as a note in the config options the crypto requirement

as a solution here?

Reply:
> I'd suggest to make the feature enabled conditionally, depending on the available python-cryptography version instead of changing the keylime python-cryptography requirement unconditionally.

I guess it's good to have the fallback mechanism to support IAK + IDevID and also EK cert, of which all the agent should provide. With older versions of python we will then always have to fall back to EK cert.

Reply:
> * Including as a note in the config options the crypto requirement

Yes, write it to the description of `tpm_identity = default`.

Reply:
Am I right in saying the requirements.txt should keep my change? Since it is only used by pip installs afaik, the packagers ignore it and just add python3-cryptography as a build requirement, which if I add the version check will not cause an issue (they just wont have access to this feature).

Reply:
Check added in [this commit.](https://github.com/keylime/keylime/pull/1477/commits/3e925f4aa2d49c49e16b74d734357112ccdf4de3)

Reply:
> Am I right in saying the requirements.txt should keep my change? Since it is only used by pip installs afaik, the packagers ignore it and just add python3-cryptography as a build requirement, which if I add the version check will not cause an issue (they just wont have access to this feature).

Better revert the change in the `requirements.txt` as it would be misleading. Having the version `38.0.0` is not a hard requirement.

Reply:
This would mean that anyone installing keylime with an old version of cryptography already installed wouldn't be able to use this. I suppose the problem is that requirements.txt is being used both as a pip installer file and as description of the project. In which case maybe the minimum requirements should be added to documentation instead and the file could be renamed to something pip related. 

At the moment these changes will work for users installing via installer.sh, and will work for rpm install with access to both new and old cryptography with the caveat that old crypto users will only have access to EK registration. Reverting will stop this working for a subset of instller.sh users. If we are OK with that I can just revert, but I just wanted to make sure as this will reduce the quantity of people that will have access to this.

Reply:
I would also revert it for now since probably not a whole lot of TPMs have IDevID and IAK support, yet, except for those used in some clouds of course.

Reply:
@ansasaki @stefanberger this has been reverted now so if you are happy with the check this can be resolved.